File: 2022/day06/typescript/solution.ts
export function partOne(input: string) {
    return findRepeatInWindow(input, 4);
}

export function partTwo(input: string) {
    return findRepeatInWindow(input, 14);
}

function findRepeatInWindow(input: string, windowSize: number) {
    for (let i = windowSize - 1; i < input.length; i++) {
        const marker = new Set<string>();

        for (let j = 0; j < windowSize; j++) {
            marker.add(input[i - j]);
        }
        if (marker.size === windowSize) return i + 1;
    }
    return -1;
}

File: 2022/day06/python/solution.py
def part_one(input):
    return find_repeat_in_window(input.strip(), 4)


def part_two(input):
    return find_repeat_in_window(input.strip(), 14)


def find_repeat_in_window(input, windowSize):
    for i in range(windowSize - 1, len(input)):
        marker = set()

        for j in range(0, windowSize):
            marker.add(input[i-j])

        if len(marker) == windowSize:
            return i + 1

    return -1


with open("../input", "r") as file:
    input = file.read()
    print("part one:", part_one(input))
    print("part two:", part_two(input))
